src/compiler/checker.ts

```diff
@@ -14542,11 +14542,15 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
     }
 
     function getResolvedApparentTypeOfMappedType(type: MappedType) {
-        const typeVariable = getHomomorphicTypeVariable(type);
-        if (typeVariable && !type.declaration.nameType) {
-            const constraint = getConstraintOfTypeParameter(typeVariable);
-            if (constraint && everyType(constraint, isArrayOrTupleType)) {
-                return instantiateType(type, prependTypeMapping(typeVariable, constraint, type.mapper));
+        const target = (type.target ?? type) as MappedType;
+        const typeVariable = getHomomorphicTypeVariable(target);
+        if (typeVariable && !target.declaration.nameType) {
+            const constraint = getConstraintTypeFromMappedType(type);
+            if (constraint.flags & TypeFlags.Index) {

```

Comment:
I wonder if this check shouldn't be replaced with an assert - it's quite a strong invariant that this branch can only be entered after `getHomomorphicTypeVariable` returns true and that requires the constraint to be an `Index` type.

Reply:
An assert here wouldn't be right. There many cases where an instantiation of a homomorphic mapped type doesn't have a `keyof XXX` constraint anymore. For example, the homomorphic `Readonly<T>` has the constraint type `keyof T`, but the instantiation `Readonly<{ a: string, b: string }>` has the constraint type `"a" | "b"`. What we're looking for here are instantiations where the constraint type is _still_ a `keyof XXX`. That's what wasn't handled before.

Reply:
Thank you for the explanation - it makes sense, `getHomomorphicTypeVariable` is called now with the `target` and not the `type` after all.